src/Microsoft.Windows.CsWin32/templates/PInvokeClassHelperMethods.cs

```diff
@@ -0,0 +1,21 @@
+﻿/// <summary>
+/// This class wrapper is stripped so that individual helper methods may be injected into the generated PInvoke class.
+/// </summary>
+internal class PInvokeClassHelperMethods
+{
+	private static void EnsureNullTerminated(Span<char> buffer, string parameterName)
+	{
+		if (buffer.Length == 0 || buffer[buffer.Length - 1] != '\0')
+		{
+			for (int i = buffer.Length - 2; i >= 0; i--)
+			{
+				if (buffer[i] == '\0')
+				{
+					return;
+				}
+			}
+
+			throw new ArgumentException("Required null terminator is missing.", parameterName);
+		}
```

Comment:
(Also, are you sure that the current two checks done outside the loop actually execute differently than if you just had the loop starting at `buffer.Length - 1`?)

Reply:
(Also, are you sure that the current two checks done outside the loop actually execute differently than if you just had the loop starting at `buffer.Length - 1`?)

Reply:
You linked the generic helper, but I think this would actually be using the char helper: https://source.dot.net/#System.Private.CoreLib/SpanHelpers.Char.cs,cd1a94968bb436e7

Reply:
I'm not calling that one. But I guess I'm not calling the one I linked to either. 

I'd be calling [this](https://source.dot.net/#System.Private.CoreLib/MemoryExtensions.cs,ec71edb0237f41ee,references) which defers to [this](https://source.dot.net/#System.Private.CoreLib/SpanHelpers.Char.cs,74035f44dfacd2ac).

So I do see vectors used there. I'm not familiar with them. Does that make it faster due to hardware acceleration?

Reply:
Yes, exactly. More on the general topic of vectorization in .NET that does refer to IndexOf: https://devblogs.microsoft.com/dotnet/regular-expression-improvements-in-dotnet-7/#vectorization